Support avx 512 instructions set

JivanH opened the pull request:
In this branch I have added support for AVX-512 instructions set.

Current version of dynamorio cannot detect difference between `bound` and `AVX-512` instructions if the opcode starts with `0x62`. This problem is solved in the request. After that `Illegal instruction` error is fixed for these instructions set.

For now I have on going problem with disassembly of newly added instructions.

Original disassemble:

> 62 f1 fd 48 6f e9       vmovdqa64 %zmm1,%zmm5

 DynamoRIO disassemble:

> 62 f1 fd 48 6f e9       vmovdqa %zmm1[16byte] -> %zmm5[16byte]

Carrotman42 commented:
Hendrik, since you were thinking through this feature I think you taking a look makes the most sense.

hgreving2304 commented:
@JivanH , thanks for the patch, we'll get back to you shortly. We may have to merge it with some other patches going into the trunk shortly. What are your short-terms plans for AVX-512? Any kind of contributions are always appreciated. Talk to you soon.

JivanH commented:
@hgreving2304 Currently I try to fix issue on disassembly. And after that support API for AVX-512 instructions (fox example INSTR_CREATE_vmovdqa32 and INSTR_CREATE_vmovdqa64)

hgreving2304 commented:
Great, could you please merge your patch.
For the evex prefix encoder, would you be able to start adding some tests?
Could you elaborate on known details possibly not supported - yet - by the evex decoder? For example the a-bits for the opmasks. Please try to document which fields you are currently covering and which not.
Thanks for working on this!

JivanH commented:
@hgreving2304 I will fix all mentioned points above in next week. Do you have some document file in which I should express which fields had covered and which not.

hgreving2304 commented:
> @hgreving2304 I will fix all mentioned points above in next week. Do you have some document file in which I should express which fields had covered and which not.

A good start is Intel Vol. 2A "2.6.1 Instruction Format and EVEX" and https://en.wikipedia.org/wiki/EVEX_prefix.

JivanH commented:
@hgreving2304  I mean where to write documentation about filed I covered? Do you have any spacial file for this or I should write it in pull request?

hgreving2304 commented:
> @hgreving2304 I mean where to write documentation about filed I covered? Do you have any spacial file for this or I should write it in pull request?

Ideally, you would make the evex encoder complete. If it's clearly structured you may want to add XXX i#1312 comments that point out where/what is missing. As well as we can discuss here.

hgreving2304 commented:
While it's usually policy not to merge an ongoing review, in this case, could you please merge and resolve conflicts, since a lot of code overlaps with recent changes on GitHub. Thanks.

JivanH commented:
@hgreving2304 I have merged master with my branch. Also have added support of parsing all fields of EVEX instructions.

hgreving2304 commented on core/arch/opnd.h:
```diff
@@ -1169,6 +1169,38 @@ extern const reg_id_t dr_reg_fixer[];
 #    define REG_YMM13 DR_REG_YMM13
 #    define REG_YMM14 DR_REG_YMM14
 #    define REG_YMM15 DR_REG_YMM15
+#    define REG_ZMM0 DR_REG_ZMM0
```
Do not add new REG_ enums, they are deprecated. Use the DR_REG_ enum instead.

hgreving2304 commented on core/arch/opnd.h:
```diff
@@ -2465,6 +2497,14 @@ DR_API
+  * \deprecated Prefer reg_is_strictly_zmm().
+  */
+bool
+reg_is_zmm(reg_id_t reg);
```
reg_is_ queries are deprecrated for simd, do not add new ones. Use reg_is_strictly_ instead.

hgreving2304 commented on core/arch/opnd_shared.c:
```diff
@@ -1986,7 +1986,7 @@ reg_is_extended(reg_id_t reg)
             (reg >= DR_REG_START_XMM + 8 && reg <= DR_REG_STOP_XMM) ||
             (reg >= DR_REG_START_YMM + 8 && reg <= DR_REG_STOP_YMM) ||
-            (reg >= DR_REG_START_ZMM + 8 && reg <= DR_REG_STOP_ZMM) ||
+            (reg >= DR_REG_START_ZMM + 16 && reg <= DR_REG_STOP_ZMM) ||
```
Should be +8, please revert.

JivanH replied:
@hgreving2304 whether **reg_is_extended()** does not checks `reg` is high register or not? For zmm registers high registers are zmm16-zmm31.

hgreving2304 replied:
reg_is_extended is used to determine < 8 registers in dimode. It's the same for AVX-512 (32-bit still has 8 registers). We may need a new function like reg_is_super_extended (bad name) for AVX-512 (but less architecture specific).

hgreving2304 commented on core/arch/opnd_shared.c:
```diff
@@ -2082,9 +2082,9 @@ reg_get_bits(reg_id_t reg)
         return (byte)((reg - DR_REG_START_XMM) % 8);
     if (reg >= DR_REG_START_YMM && reg <= DR_REG_STOP_YMM)
-        return (byte)((reg - DR_REG_START_YMM) % 8);
+        return (byte)((reg - REG_START_YMM) % 8);
```
See above, the REG_ enum/defines are deprecated. Use DR_REG_ version, please revert.

hgreving2304 commented on core/arch/opnd_shared.c:
```diff
@@ -2082,9 +2082,9 @@ reg_get_bits(reg_id_t reg)
+        return (byte)((reg - REG_START_YMM) % 8);
     if (reg >= DR_REG_START_ZMM && reg <= DR_REG_STOP_ZMM)
-        return (byte)((reg - DR_REG_START_ZMM) % 8);
+        return (byte)((reg - DR_REG_START_ZMM) % 16);
```
Revert. Function is for modrm bits only.

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -708,6 +708,135 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+* the first opcode byte.
+*/
+static byte *
+read_evex(byte *pc, decode_info_t *di, byte instr_byte,
```
Will review shorty - further review pending.

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -708,6 +708,135 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+    CLIENT_ASSERT(info->type == EVEX_PREFIX_EXT, "internal evex decoding error");
+
+    /* If 64-bit mode, P[3:2] are 0 and P[10] is 1 then this is evex */
+    if (X64_MODE(di) && !TEST(0xC, *pc) && TEST(0x04, *(pc + 1))) {
```
Should be similar to VEX. Either di mode (0x62 is always EVEX) or (||) test modrm MOD for 0b11, which is invalid for bound and therefore EVEX.

The bits P[3:2] and P[10] should be 0b00 and 0b1 respectively, otherwise return #UD similar as read_vex does it.

JivanH replied:
Please resolve conversations if my changes meet requirements.

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -708,6 +708,135 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+        info = &evex_prefix_extensions[0][1];
+    } else {
+
+        CLIENT_ASSERT(!X64_MODE(di), "internal evex decoding error");
```
It's not an error, it would be is_evex = false, ret_info = info, return pc;

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -708,6 +708,135 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+    if (!TEST(0x80, prefix_byte)) {
+        di->prefixes |= PREFIX_REX_R;
+    }
+
```
nit: compact, remove empty lines.

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -1191,7 +1323,8 @@ decode_reg(decode_reg_t which_reg, decode_info_t *di, byte optype, opnd_size_t o
         reg = di->reg;
-        extend = X64_MODE(di) && TEST(PREFIX_REX_R, di->prefixes);
+        extend = X64_MODE(di) && (TEST(PREFIX_REX_R, di->prefixes) ||
+                 TEST(PREFIX_EVEX_RR, di->prefixes)); /* for evex instructions */;
```
Would need another, different 'extend' for EVEX R'? I think the meanings of R and R' overlap in 'extend' otherwise.

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -708,6 +708,135 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+    }
+
+    evex_pp = prefix_byte & 0x03;
+    di->vex_vvvv = (prefix_byte & 0x78) >> 3;
```
Rename to vvvv? Or better: make a union with evex_vvvv, so other code like operand decoder can later refer to by an unique name. This implies changes further below in this patch, I think the AVX-512 instructions need new table entries with new operand types (i.e. TYPE_B into TYPE_BB, unfortunately there are no predefined type name in the Intel manuals). I haven't thought this through, but can we use existing instruction tables with just wider operands? That can't be true, since there are now mask operands, i.e. different, more operands, etc.

JivanH replied:
vex_vvvv and evex_vvvv have the same logic. I not understand why  rename vex_vvvv?
I agree with "AVX-512 instructions need new table entries with new operand types" point. I will add that table in my future pull requests, when I will support disassemble of AVX-512 instructions.

hgreving2304 commented:
>  I will add that table in my future pull requests, when I will support disassemble of AVX-512 instructions.

Thanks. All pull requests will need tests, and tests will need the (or some-) tables. Are you able to share your plans with us at this point? Reason I am asking is that there is an effort underway to add AVX-512 to DynamoRIO here at Google. The scalar VEX encoded opmask instructions will be added shortly, next are the EVEX versions. Your efforts will need to be coordinated with us in some way. Your EVEX decoder could be ok after implementing the review comments. Are you able to add some instructions to the tables? If so, which ones? A good candidate are the vmov* AVX-512 EVEX instructions. An alternative approach is for us to add the EVEX decoder (similar or the same as yours), add some basic tables and tests from our side, and you could help adding instructions to the table then. Any contributions are always appreciated. 

xref #1312

JivanH commented:
@hgreving2304 I agree with you. My pull request needs tests. I will add vmov* AVX-512 EVEX instructions  to the tables. Do you have defined type for operand with writemask {k1}{z} ? For example vmovdqa32 ymm1 {k1}{z}, ymm2/m256 instruction.

hgreving2304 commented:
> @hgreving2304 I agree with you. My pull request needs tests. I will add vmov* AVX-512 EVEX instructions to the tables. Do you have defined type for operand with writemask {k1}{z} ? For example vmovdqa32 ymm1 {k1}{z}, ymm2/m256 instruction.

Would need more details about your plans + time. Work on EVEX prob. starts next week.
There will be a patch going into upstream by next week that will define a TYPE_K_EVEX that stands for EVEX.aaa encoded opmask registers. Go with that for now. Haven't thought it through completely but the size may not have to be variable (e.g. the decoding table can have a fixed size Kevw, Kevb, etc.) because the size should be implicit as defined by the opcode.

JivanH commented:
I am waiting for TYPE_K_EVEX patch.  I suppose support of vmov* instruction will take maximum 2-3 days.

hgreving2304 commented:
> I am waiting for TYPE_K_EVEX patch. I suppose support of vmov* instruction will take maximum 2-3 days.

There will be

#define KEb TYPE_K_EVEX, OPSZ_1
#define KEw TYPE_K_EVEX, OPSZ_2
#define KEd TYPE_K_EVEX, OPSZ_4
#define KEq TYPE_K_EVEX, OPSZ_8

in decode_table.c, with TYPE_K_EVEX meaning EVEX.aaa encoded k register.

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -1223,14 +1347,21 @@ decode_reg(decode_reg_t which_reg, decode_info_t *di, byte optype, opnd_size_t o
-                    : (extend ? (REG_START_XMM + 8 + reg) : (REG_START_XMM + reg)));
+      return (
+          TEST(PREFIX_EVEX_RR, di->prefixes)
+              ? (extend ? (DR_REG_START_ZMM + 16 + reg)
```
Missing simd > 24. Let's don't mix the meaning of 'extend' or 'extended' being > 8 registers (rex.R) and the extension of AVX-512 from 16 to 32 simd registers anywhere in core/.

AssadHashmi commented:
Can one of the admins verify this patch?

JivanH commented:
@hgreving2304 I have added support of some instructions. May you check my changes? Currently I working on adding tests for my changes and fixing failed tests.

Supported instructions:
  vaddps, vaddss, vaddpd, vaddsd
  vsubps, vsubss, vsubpd, vsubsd
  vmulps, vmulss, vmulpd, vmulsd
  vdivps, vdivss, vdivpd, vdivsd
  vminps, vminss, vminpd, vminsd
  vmaxps, vmaxss, vmaxpd, vmaxsd

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -701,6 +701,118 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+            return pc;
+        }
+
+        info = &evex_prefix_extensions[0][1];
```
*is_evex = true missing?

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -701,6 +701,118 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+        info = &evex_prefix_extensions[0][1];
+    } else {
+        /* not evex */
+        is_evex = false;
```
*is_evex

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -711,8 +823,11 @@ read_prefix_ext(const instr_info_t *info, decode_info_t *di)
     int code = (int)info->code;
     int idx = (di->rep_prefix ? 1 : (di->data_prefix ? 2 : (di->repne_prefix ? 3 : 0)));
-    if (di->vex_encoded)
+    if (di->vex_encoded || di->evex_encoded) {
```
So we're mixing vex and evex entries in the prefix_ext tables? The 'else' part is currently dead. I think the "|| di->evex_encoded" is a bug and you intended to add 4 entries to the prefix_ext tables for evex?

JivanH replied:
Yes. I have added 4 entries to the prefix_ext tables for evex.

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -1226,19 +1354,25 @@ decode_reg(decode_reg_t which_reg, decode_info_t *di, byte optype, opnd_size_t o
-                    ? (extend ? (REG_START_YMM + 8 + reg) : (REG_START_YMM + reg))
-                    : (extend ? (REG_START_XMM + 8 + reg) : (REG_START_XMM + reg)));
+    case TYPE_VSIB: {
+      byte extend_reg = extend ? reg + 8 : reg;
```
I think there should be another additional notion of extended for the 16 new upper registers of AVX-512 in some form or the other, mentioned in previous comment. How are the upper 16 registers currently encoded?

hgreving2304 commented on core/arch/x86/decode_private.h:
```diff
@@ -414,6 +430,7 @@ enum {
     TYPE_K_REG,                  /* modrm.reg selects k0-k7 */
     TYPE_K_VEX,                  /* vex.vvvv field selects k0-k7 */
     TYPE_K_EVEX,                 /* evex.aaa field selects k0-k7 */
+    TYPE_K_EVEX_V,               /* reg of modrm selects xmm/ymm/zmm with opmask */
```
Type "K" is for opmask only. modrm reg field for xmm/ymm already exists, called type "V". Even though this is not officially spelled out like this in the Intel Specs., this is a natural extension of type "V".

JivanH replied:
You are right the modrm reg field already exists. This field used for representing zmm register with opmask for example zmm1{K1}{z}.

hgreving2304 replied:
Each type is for a distinct interpretation of one bit-field. zmm1{K1}{z} implies 3 fields (one for zmm, one for k1, another bit for 'z') if needed. You'd add a type if there is no one available yet. In this case, everything is already there: For example modrm.reg with EVEX.RR' for zmm (see each individual instruction in Intel's Spec.). k zeroing or merge-mask should (almost-?) always be TYPE_K_EVEX (EVEX.aaa). There is probably no type needed for {z} (merge-mask bit).

JivanH replied:
TYPE_K_EVEX_V  is merge of TYPE_K_EVEX and TYPE_V. For {z} field there is no meter to add type defination. With this way I can decide wheter operand of instruction has opmask or not?
Whether I understand to you correctly? You suggest to instead of defining a new type just use TYPE_V and add K* registers decoding evex.aaa.

hgreving2304 replied:
There need to be a new size _evex64 types like the existing ones for _vex32 that sould be used in combination with TYPE_V. TYPE_K is for mask registers only, there's no mixup of different register types in one type - yet. I think you're running into this because EVEX adds another mask operand. We might have to add another column to instr_info_t. Any other ideas?

JivanH replied:
I think adding new column in instr_info_t is not good idea, because it will bring extra memory consumption (most of instructions has not opmask field). 
I  suggest two versions.
1) Because of destination operand of instructions can be zmm1{k1} (pair of modrm register and opmask register) then I suggest to define a new type which will represent combination of these registers, for example TYPE_K_EVEX_V. 
2) We may use types of VEX instructions and durring disassemble add k* register if evex.aaa was set (z field too). For  INSTR_CREATE_* we should add new iterface which will contain info about opmask. For example vmovupd instruction will have this defination
 #define INSTR_CREATE_vmovupd(dc, d, s) instr_create_1dst_k_1src((dc), OP_vmovupd, (d), (k), (s))

hgreving2304 replied:
Having merged types like TYPE_K_EVEX_V is not the way to go, the type is distinct for one bit-field. It would either be a new field in instr_info_t, or some kind of pointer to new table entry like pusha does it. Need to get quantified how many instructions there are with > 3 sources.

hgreving2304 replied:
Jivan, could you give us an idea by any chance who're you're working with / for, your long-term plans etc. Asking because as mentioned before, there will be / is considerable development for AVX-512 now at Google, these basic things here will overlap significantly. Once these basic things will be done, work can be more easily split up upstream, e.g. adding new instructions to the tables.

derekbruening replied:
There is an existing mechanism used to support instructions with more than 3 sources or more than 2 destinations: the extra_operands[] table holds a further entry (or 2 further entries if needed).  If we're adding a smallish number of such many-source instructions we should use that existing mechanism.  If we're talking about hundreds then maybe we would consider a new way to do it if we feel the extra entry is too awkward or inefficient: are there hundreds of AVX-512 instructions with 4 or more sources?

hgreving2304 replied:
A first scan through the Specs looks like that 3 sources should mostly be enough.

JivanH replied:
@hgreving2304 I also think thant 3 sources and 2 destinations should mostly be enough.
For zmm1{k1} case may we use the first destination of instr_info_t for zmm1 and the second destination for k1 registers? If "yes" then I think we not need to define a new type. We can just use TYPE_V for the first destination (zmm*)  and TYPE_K_EVEX for the second destination (k*).

JivanH replied:
@derekbruening Thank you for your suggestion. How I saw most part of insteuctions have less than 4 sources.

hgreving2304 replied:
@JivanH yes, except the mask register is not a destination, hence needs to be a source. Another friendly reminder to synch with us about your future near term plans and your background in more detail. Reason again is that here is R&D at Google that will otherwise likely merge patches coming soon that in part will resemble what you're doing (in addition to various other missing pieces in AVX-512). We're open to splitting up in particular the various instruction table extensions but this needs to be well coordinated. Also glad to have a phone conversation about it.

JivanH replied:
Hendrik I work in ISP RAS. I going on to contribute with you for a month (for this patch). During that period of time I will add support for insteuctions with 0x0F implied leading opcode bytes. Of course the most part of time will take constructing architectire. Adding support/info of new insteuctions in tables is needs less effort. (This is my personaly opinion)

hgreving2304 commented on core/arch/x86/decode_table.c:
```diff
@@ -1483,6 +1483,19 @@ const instr_info_t * const op_instr[] =
 #define KEw TYPE_K_EVEX, OPSZ_2
 #define KEd TYPE_K_EVEX, OPSZ_4
 #define KEq TYPE_K_EVEX, OPSZ_8
+#define KEVps TYPE_K_EVEX_V, OPSZ_16_32_evex64
```
What's the rationale behind this? There shouldn't be any ps/pd in the operand type + size combo. The ps/pd relates to element size and not to operand size (e.g. the 64-byte vector is still a 64-byte vector, no matter whether ps/pd.

JivanH replied:
Ok. I will change with OPSZ_64.

hgreving2304 replied:
> There shouldn't be any ps/pd in the operand type + size combo.
Clarification: w.r.t. mask register. Actually TBD, not sure right now. but..

.. the KE* types are the mask operand types. Specifically the EVEX.aaa mask register, which is 8 bytes max. (There is no OPSZ_64 opmask.)

JivanH replied:
@hgreving2304 I understood you. With changing size of operand  I going on to change type name too. In above comment I described why type was named it this way.

hgreving2304 commented on core/arch/x86/decode_table.c:
```diff
@@ -2783,6 +2807,10 @@ const instr_info_t prefix_extensions[][8] = {
     {OP_vmovsldup,0xf30f1210, "vmovsldup", Vvs, xx, Wvs, xx, xx, mrm|vex, x, END_LIST},
     {OP_vmovlpd,  0x660f1210, "vmovlpd", Vq_dq, xx, Hq_dq, Mq, xx, mrm|vex, x, tpe[3][6]},
     {OP_vmovddup, 0xf20f1210, "vmovddup", Vvd, xx, Wvq_dq, xx, xx, mrm|vex, x, END_LIST},
+    {INVALID, 0x0f1210, "vmovlps", xx, xx, xx, xx, xx, no, x, NA},
```
INVALID?

JivanH replied:
Yes, because  not implemented yet. Names was written only for helping purpose.

JivanH replied:
@hgreving2304 please resolve discussion if I have answered your question.

hgreving2304 replied:
As long as it is INVALID, it need to be name "(bad)". Any _valid_ instruction needs to be accompanied by a INSTR_CREATE test in ir_x86.c and preferably by another decoder test (there will be a suite going upstream for this soon).

hgreving2304 commented on core/arch/x86/decode_table.c:
```diff
@@ -4011,6 +4475,10 @@ const instr_info_t prefix_extensions[][8] = {
     {OP_vmovdqa, 0x660f6f10, "vmovdqa", Vx, xx, Wx, xx, xx, mrm|vex, x, tpe[113][6]},
     {INVALID,   0xf20f6f10, "(bad)", xx, xx, xx, xx, xx, no, x, NA},
+    {INVALID, 0x0f6f10, "(bad)", xx, xx, xx, xx, xx, no, x, NA},
+    {INVALID, 0xf306f10, "vmovdqu32/64", xx, xx, xx, xx, xx, no, x, NA}, /* W0/W1 */
```
u32, u64 are 2 different opcodes in DR. There a distinct based on VEX_W (see Intel Specs)

JivanH replied:
You are right. They have different opcodes. They differs only in EVEX.W bit. I wrote in this style only for hint about that  entry. That will be changed.

JivanH replied:
@hgreving2304 please resolve discussion if I have answered your question.

JivanH commented:
https://github.com/DynamoRIO/dynamorio/pull/3512#pullrequestreview-231785154
Thank you for your review and suggestions. Of course I will test instruction whichs I was added.

AssadHashmi commented:
run arm tests

hgreving2304 commented on core/arch/decode.h:
```diff
@@ -264,6 +264,7 @@ enum {
     OPSZ_8_rex16_short4,  /**< Intel 'v' * 2 (far call/ret) */
     OPSZ_12_rex40_short6, /**< unresolved iret */
     OPSZ_16_vex32,        /**< 16 or 32 bytes depending on VEX.L (AMD/Intel 'x'). */
+    OPSZ_16_32_evex64,    /**< 16, 32 or 64 bytes depending on EVEX.LL'. */
```
Needs to be added at the end of the enum (binary compatibility)

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -701,6 +701,119 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+    info = *ret_info;
+
+    CLIENT_ASSERT(info->type == EVEX_PREFIX_EXT, "internal evex decoding error");
+    /* If 32-bit mode and mod selects for memory, this is not vex */
```
s/vex/evex/

hgreving2304 commented on core/arch/x86/decode.c:
```diff
@@ -701,6 +701,119 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
+
+    byte evex_mm;
+    CLIENT_ASSERT(info->type == PREFIX, "internal evex decoding error");
+    /* fields are: R, X, B, R', 00, mm.  R, X, B and R' are inverted. */
```
Could you mention in a comment that Intel fails to document this in "Vol-2A 2.6 INTEL® AVX-512 ENCODING" but we tested this with common assemblers, as well as the patent WO2012134532A1 mentions that the bits are in fact inverted the same way as in the VEX prefix, in order to make it distinct from the bound instruction in 32-bit mode.

hgreving2304 commented:
@JivanH , few things are missing in this patch / won't work like this. Examples are,

* Instructions will have to be multiplexed through evex_extensions, most likely shared w/ vex_extensions,
  i.e. e_vex_extensions.
* Most or all of them will have to go through evex_w_extensions.
* Most of them will have to go through something like evex_ll_extensions (L'), prob. need to ask a 
  read_ll_prefix or alike.
* Proper handling of register size extensions w/ L bits (along the same lines as above, mentioned in 
  comment above).
* No mixup of K types and zmm types. Quantifying the need for 3 > sources (although we did that
  and it looks it is ok).
* Various macro tests and testing needs to be added.

Sorry that this work here overlaps with work at Google, I'm suggesting you wait until a basic patch will go upstream soon that you could use as a template to help adding more instructions. I hope you're interested doing that. For bigger things like that, please always share your plans with dynamorio-devs <dynamorio-devs@googlegroups.com> ahead of time.

If you like to help, also need to even more precisely define the instructions you like to add, please refer to i#1312. There are many other AVX-512 things left TODO as well, e.g.
* Proper context switching for AVX-512.
* Lazy context AVX-512 context switching (can share more info if needed).
* AVX-512 feature bit recognition in proc/ for AVX512BW, DQ, etc.
* API extensions like providing rounding modes etc.
* Also things like i#2985 and extending it to AVX-512.
* Handling of different types of compressed displacement.

JivanH commented:
> @JivanH , few things are missing in this patch / won't work like this. Examples are,
> 
> * Most instructions will have to go through evex_w_extensions.
> * Some of them might have to go through something like evex_ll_extensions (L'), prob. need to ask a
>   read_ll_prefix or alike.
> * Some instructions will have to be multiplexed through evex_extensions, most likely shared w/ vex_extensions,
>   i.e. e_vex_extensions.
> * Proper handling of register size extensions w/ L bits (along the same lines as above, mentioned in
>   comment above). Mainly "Bit X now expands r/m along with bit B when the SIB byte is not present, which allows 32 SIMD registers." and "Bit R’ for expanding reg, which allows for 32 SIMD registers.".
> * No mixup of K types and zmm types. Quantifying the need for 3 > sources (although we did that
>   and it looks it is ok).
> * Encoder part (e.g. requires_evex / W bit)
> * Various macro tests and testing needs to be added.

@hgreving2304 I have added evex_extensions and evex_W_extension tables. Have not added evex_ll_extensions table because cannot find some instruction for that table. If you know instruction let me know. Split types K and type zmms. Added vmovdqa32, vmovdqa64, vmovdqu8, vmovdqu16, vmovdqu32, vmovdqu64 and corresponding  INSTR_CREATE_*. I think for fully supporting these instructions remains only tests.

hgreving2304 commented:
Getting back to you shortly, we might merge parts of this PR first. Thanks for your work.

hgreving2304 commented:
@JivanH, could you please set up a pull request that only includes your read_evex function. Please add a
#if /* DISABLED_UNTIL_BUG_1312_IS_FIXED
#endif
around it in order to avoid an unused warning error. There will be a small patch going in soon using your read_evex and doing one api.ir test's roundtrip that you're welcome to use as a template. Please do that asap.

Alternatively, I can merge parts of above into another pull request, but I'd prefer to give you credit in the committer's history. Yet another alternative is for us to edit this pull request here heavily an merge that. Please get back to me asap. Thanks in advance.

JivanH commented:
> #if /* DISABLED_UNTIL_BUG_1312_IS_FIXED
> #endif

@hgreving2304 I have set up a new pull request. Please check it.

JivanH commented:
run arm tests

hgreving2304 commented:
Now is a good time to merge w/ upstream. For any instruction you're adding, please add a complete roundtrip test to the x86 api.ir test as it has been done in upstream. Please mark instructions you intend to work on and intend to complete with (WIP) in https://github.com/DynamoRIO/dynamorio/issues/1312. Instructions added to upstream are (Done).  Consider enabling subtests in decenc.drdecode_decenc_x86(_64) if subtest passes. At any time, feel free to open a conversation with me or on dynamorio-devs@googlegroups.com for clarifications. There may be corner cases with operand sizes, type names, disassembler etc. Any question or discussion is welcome. Thanks!

AssadHashmi commented:
run arm tests

hgreving2304 commented on core/arch/x86/opcode.h:
```diff
@@ -1318,6 +1318,12 @@ enum {
     /* 1155 */ OP_ktestb,   /**< IA-32/AMD64 AVX-512 ktestd opcode. */
     /* 1156 */ OP_ktestq,   /**< IA-32/AMD64 AVX-512 ktestd opcode. */
     /* 1157 */ OP_ktestd,   /**< IA-32/AMD64 AVX-512 ktestd opcode. */
+    /* 1158 */ OP_vmovdqa32, /**< IA-32/AMD64 AVX-512 ktestd opcode. */
```
Exactly these instructions will go upstream shortly, along with vmovss, vmovsd. Please pick instructions from i#1312, mark with (WIP), maybe (WIP JivanH) is better... There are still hundreds of instructions left, plenty of work to pick from!

AssadHashmi commented:
run arm tests
